src/passes/MemoryPacking.cpp

```diff
@@ -23,151 +38,607 @@
+struct Range {
+  bool isZero;
+  size_t start;
+  size_t end;
+};
+
+// A function that produces the transformed bulk memory op
+using Replacement = std::function<Expression*(Function*)>;
+
+// Maps each bulk memory op to the replacement that must be applied to it.
+using Replacements = std::unordered_map<Expression*, Replacement>;
+
+// memory.fill: 2 byte opcode + ~2 immediate bytes + ~(3*2) byte operands
+const size_t MEMORY_FILL_SIZE = 10;
+
+// data.drop: 2 byte opcode + ~1 byte index immediate
+const size_t DATA_DROP_SIZE = 3;
+
+namespace {
+
+Expression* makeMemorySize(Builder& builder) {
+  return builder.makeBinary(ShlInt32,
+                            builder.makeHost(MemorySize, Name(), {}),
+                            builder.makeConst(Literal(int32_t(16))));
+}
+
+} // anonymous namespace
 
 struct MemoryPacking : public Pass {
-  bool modifiesBinaryenIR() override { return false; }
+  size_t dropStateGlobalCount = 0;
 
-  void run(PassRunner* runner, Module* module) override {
-    if (!module->memory.exists) {
-      return;
+  void run(PassRunner* runner, Module* module) override;
+  void optimizeBulkMemoryOps(PassRunner* runner, Module* module);
+  void getSegmentReferrers(Module* module,
+                           std::vector<std::vector<Expression*>>& referrers);
+  void dropUnusedSegments(std::vector<Memory::Segment>& segments,
+                          std::vector<std::vector<Expression*>>& referrers);
+  bool canSplit(const Memory::Segment& segment,
+                const std::vector<Expression*>& referrers);
+  void calculateRanges(const Memory::Segment& segment,
+                       const std::vector<Expression*>& referrers,
+                       std::vector<Range>& ranges);
+  void createSplitSegments(Builder& builder,
+                           const Memory::Segment& segment,
+                           std::vector<Range>& ranges,
+                           std::vector<Memory::Segment>& packed,
+                           size_t segmentsRemaining);
+  void createReplacements(Module* module,
+                          const std::vector<Range>& ranges,
+                          const std::vector<Expression*>& referrers,
+                          Replacements& replacements,
+                          const Index segmentIndex);
+  void replaceBulkMemoryOps(PassRunner* runner,
+                            Module* module,
+                            Replacements& replacements);
+};
+
+void MemoryPacking::run(PassRunner* runner, Module* module) {
+  if (!module->memory.exists) {
+    return;
+  }
+
+  auto& segments = module->memory.segments;
+
+  // For each segment, a list of bulk memory instructions that refer to it
+  std::vector<std::vector<Expression*>> referrers(segments.size());
+
+  if (module->features.hasBulkMemory()) {
+    // Remove any references to active segments from bulk memory
+    // operations. This means we only need to track replacements for passive
+    // segments.
```

Comment:
I rewrote the comment to go into more details. It now reads,
```
 127     // Optimize out memory.inits and data.drops that can be entirely replaced
 128     // with other instruction sequences. This can increase the number of unused
 129     // segments that can be dropped entirely and allows later replacement
 130     // creation to make more assumptions about what these instructions will look
 131     // like, such as memory.inits not having both zero offset and size.
```